core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -148,12 +149,32 @@ private[spark] class CoarseGrainedExecutorBackend(
    * executor exits differently. For e.g. when an executor goes down,
    * back-end may not want to take the parent process down.
    */
-  protected def exitExecutor(code: Int, reason: String, throwable: Throwable = null) = {
+  protected def exitExecutor(code: Int,
+                             reason: String,
+                             throwable: Throwable = null,
+                             notifyDriver: Boolean = true) = {
+    val message = "Executor self-exiting due to : " + reason
     if (throwable != null) {
-      logError(reason, throwable)
+      logError(message, throwable)
     } else {
-      logError(reason)
+      logError(message)
     }
+
+    if (notifyDriver) {
+      logInfo(s"Notifying the driver before exiting the executor")
+      rpcEnv.asyncSetupEndpointRefByURI(driverUrl).flatMap { ref =>
```

Comment:
not need to connect to driver if `driver` is None. It's usually because we cannot connect to driver.
```scala
if (notifyDriver && driver.nonEmpty) {
  driver.get.ask[Boolean](RemoveExecutor(executorId, new ExecutorLossReason(reason))).onFailure { case e =>
      logWarning(s"Unable to notify the driver due to " + e.getMessage, e)
    }(ThreadUtils.sameThread)
}
```